src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,370 @@
+	}
+	appendStringInfo(splitQuery, "%d], %s)", shardinfo->nodeid, quote_literal_cstr(shardSplitMode));
+
+	return splitQuery;
+}
+
+
+/*
+ * It creates a background job for citus_split_shard_by_split_points and executes it in background.
+ */
+void
+ExecuteSplitBackgroundJob(int64 jobid, ShardInfo shardinfo, List *SplitPoints , char* shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+	splitQuery = GetShardSplitQuery(shardinfo, SplitPoints , shardSplitMode);
+	ereport(LOG, (errmsg(splitQuery->data)));
+	int32 nodesInvolved[1];
+	nodesInvolved[0] = shardinfo->nodeid;
+	Oid superUserId = CitusExtensionOwner();
+	BackgroundTask *task = ScheduleBackgroundTask(jobid, superUserId, splitQuery->data, 0,
+												  NULL, 1, nodesInvolved);
+}
+
+
+/*
+ * It executes a query to find the average hash value in a shard considering rows with a limit of 10GB .
+ * If there exists a hash value it is returned otherwise shardminvalue-1 is returned.
+ */
+int64
+ExecuteAvgHashQuery(ShardInfo shardinfo)
+{
+	StringInfo AvgHashQuery = makeStringInfo();
+	appendStringInfo(AvgHashQuery, "SELECT avg(h)::int,count(*)"
+								   " FROM (SELECT worker_hash(%s) h FROM %s TABLESAMPLE SYSTEM(least(10, 100*10000000000/citus_total_relation_size(%s)))"
+								   " WHERE worker_hash(%s)>=%ld AND worker_hash(%s)<=%ld) s",
+					 shardinfo->distributionColumn, shardinfo->tablename,
+					 quote_literal_cstr(shardinfo->tablename),
+					 shardinfo->distributionColumn, shardinfo->shardminvalue,
+					 shardinfo->distributionColumn, shardinfo->shardmaxvalue
+					 );
+	ereport(LOG, errmsg("%s", AvgHashQuery->data));
+	SPI_connect();
+	SPI_exec(AvgHashQuery->data, 0);
+	SPITupleTable *tupletable = SPI_tuptable;
+	HeapTuple tuple = tupletable->vals[0];
+	bool isnull;
+	Datum average = SPI_getbinval(tuple, tupletable->tupdesc, 1, &isnull);
+	int64 IsResultNull = 1;
+	if (!isnull)
+	{
+		IsResultNull = 0;
+	}
+	SPI_freetuptable(tupletable);
+	SPI_finish();
+
+	if (IsResultNull == 0)
+	{
+		return DatumGetInt64(average);
+	}
+	else
+	{
+		return shardinfo->shardminvalue - 1;
+	}
+}
+
+
+/*
+ * This function executes a query and then decides whether a shard is subjected for isolation or average hash 2 way split.
+ * If a tenant is found splitpoints for isolation is returned otherwise average hash value is returned.
+ */
+List *
+FindShardSplitPoints(ShardInfo shardinfo)
+{
+	StringInfo CommonValueQuery = makeStringInfo();
+
+	/*
+	 * The inner query for extracting the tenant value having frequency > 0.3 is executed on
+	 * every shard of the table and outer query gives the shardid and tenant values as the
+	 * output.
+	 */
+	appendStringInfo(CommonValueQuery,
+					 "SELECT shardid , unnest(result::%s[]) from run_command_on_shards(%s,$$SELECT array_agg(val)"
+					 " FROM pg_stats s , unnest(most_common_vals::text::%s[],most_common_freqs) as res(val,freq)"
+					 " WHERE tablename = %s AND attname = %s AND freq > 0.2 $$)"
```

Comment:
should probably make the 0.2 configurable

Reply:
made it as citus.tenant_freq GUC